src/scheduler/Executor.cpp

```diff
@@ -79,99 +119,116 @@ void Executor::executeTasks(std::vector<int> msgIdxs,
     // Iterate through and invoke tasks
     for (int msgIdx : msgIdxs) {
         const faabric::Message& msg = req->messages().at(msgIdx);
-        int threadPoolIdx = msg.appindex() % threadPoolSize;
+
+        // If executing threads, we must always keep thread pool index zero
+        // free, as this may be executing the function that spawned them
+        int threadPoolIdx;
+        if (isThreads) {
+            assert(threadPoolSize > 1);
+            threadPoolIdx = (msg.appindex() % (threadPoolSize - 1)) + 1;
+        } else {
+            threadPoolIdx = msg.appindex() % threadPoolSize;
+        }
 
         // Enqueue the task
+        logger->trace(
+          "Assigning app index {} to thread {}", msg.appindex(), threadPoolIdx);
         threadQueues[threadPoolIdx].enqueue(std::make_pair(msgIdx, req));
 
         // Lazily create the thread
-        if (threadPoolThreads.count(threadPoolIdx) == 0) {
```

Comment:
Not much has changed here, but I moved the guts of the thread to its own function below. This makes it much easier to read and avoids this deeply indented code block.